components/base-rest-api/packages/services/lib/token-swapper-service.js

```diff
@@ -0,0 +1,85 @@
+import _ from 'lodash';
+import { Service } from '@aws-ee/base-services-container';
+import jwtDecode from 'jwt-decode';
+import { newInvoker } from './authentication-providers/helpers/invoker';
+
+const settingKeys = {
+  dbValidTokens: 'dbValidTokens',
+};
+class TokenSwapperService extends Service {
+  constructor() {
+    super();
+    this.boom.extend(['invalidToken', 403]);
+    this.dependency(['dbService', 'authenticationProviderConfigService']);
+  }
+
+  async init() {
+    this.invoke = newInvoker(this.container.find.bind(this.container));
+  }
+
+  async swap({ token, uid }) {
+    const dbService = await this.service('dbService');
+    const dbValidTokens = this.settings.get(settingKeys.dbValidTokens);
+
+    const { isSameToken } = await this.revokeValidToken({ token, uid });
+    if (isSameToken) {
+      return;
+    }
+    const record = { uid, token };
+    await dbService.helper
+      .updater()
+      .table(dbValidTokens)
+      .key({ uid })
+      .item(record)
+      .update();
+  }
+
+  async revokeValidToken({ token, uid }) {
+    const dbService = await this.service('dbService');
+    const authenticationProviderConfigService = await this.service('authenticationProviderConfigService');
+    const dbValidTokens = this.settings.get(settingKeys.dbValidTokens);
+
+    const record = await dbService.helper
+      .getter()
+      .table(dbValidTokens)
+      .key({ uid })
+      .get();
+
+    if (!record) {
+      return { isSameToken: false };
+    }
+
+    const oldToken = record.token;
+    const isSameToken = token === oldToken;
+    if (isSameToken) {
+      return { isSameToken: true };
+    }
+
+    let payload;
+    try {
+      payload = jwtDecode(oldToken);
```

Comment:
If we get into this case, it looks like the user becomes locked out. Do we want something here to clean up the database record?

Reply:
I think we discussed updating documentation for this outcome, but I did not see an update.